src/goto-instrument/synthesizer/synthesizer_utils.cpp

```diff
@@ -0,0 +1,79 @@
+/*******************************************************************\
+
+Module: Utility functions for loop invariant synthesizer.
+
+Author: Qinheping Hu
+
+\*******************************************************************/
+
+#include "synthesizer_utils.h"
+
+goto_programt::targett get_loop_head_or_end(
+  const unsigned int target_loop_number,
+  goto_functiont &function,
+  bool finding_head)
+{
+  natural_loops_mutablet natural_loops(function.body);
+
+  // iterate over all natural loops to find the loop with `target_loop_number`
+  for(const auto &loop_p : natural_loops.loop_map)
+  {
+    const goto_programt::targett loop_head = loop_p.first;
+    goto_programt::targett loop_end = loop_p.first;
+    const loopt &loop = loop_p.second;
+    for(const auto &t : loop)
+    {
+      // an instruction is the loop end if it is a goto instruction
+      // and it jumps backward to the loop head
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_end->location_number)
+        loop_end = t;
+    }
```

Comment:
We have a very similar snippet in the implementation of the `apply` option for loop contracts, to identify the `loop_end`:
https://github.com/diffblue/cbmc/blob/b466503080639f98af3329e9e9bb05ac847a2fc3/src/goto-instrument/contracts/contracts.cpp#L1085-L1101
It would be a good idea to refactor this snippet out into some util function that can be called at both locations.

Also see my comment about using the `loop_head` itself as the key for the `invariant_map`, so we wouldn't need to do a linear scan over all loops to find the right index and grab its head.

Reply:
It is a good idea. I will open another PR for it.